// angular-unit-test-generator/dependencies.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var dependency_1 = require("./constants/dependency");
var DependenciesService = (function () {
    function DependenciesService(fileContent) {
        this.fileContent = fileContent;
        this.list = this.getDependencies();
    }
    DependenciesService.prototype.getDependencies = function () {
        var _this = this;
        var constructor = this.getConstructor();
        var lineDepsMatch = constructor.match(/private (.*):\s([\w|<|>]*)/gm);
        return lineDepsMatch.reduce(function (result, lineDepMatchItem) {
            result.push(_this.generateDependency(lineDepMatchItem));
            return result;
        }, []);
    };
    DependenciesService.prototype.getConstructor = function () {
        return this.fileContent.match(/constructor\s*\([\s\S]*\)\s*{\s*}/gm)[0];
    };
    DependenciesService.prototype.generateDependency = function (lineDepMatchItem) {
        var depMatchItem = lineDepMatchItem.match(/private\s(.*):\s(.*),?/);
        var depName = depMatchItem[1];
        var depType = depMatchItem[2];
        var genericMatch = depType.match(/<(.*?)>/);
        var typeWithoutGenericMatch = depType.match(/(.*)<|(.*)/);
        return new dependency_1.Dependency({
            name: depName,
            type: depType,
            generic: genericMatch ? genericMatch[1] : undefined,
            typeWithoutGeneric: typeWithoutGenericMatch ? typeWithoutGenericMatch[1] : depType
        });
    };
    return DependenciesService;
}());
exports.DependenciesService = DependenciesService;
