// angular-unit-test-generator/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var fs = require("fs");
var util_1 = require("util");
exports.readFile = util_1.promisify(fs.readFile);
exports.writeFile = util_1.promisify(fs.writeFile);
exports.unlink = util_1.promisify(fs.unlink);
function escapeRegExpChar(text) {
    return text.replace(/\$/g, '\\$');
}
exports.escapeRegExpChar = escapeRegExpChar;
function removeExtension(filename) {
    var pointExtensionIndex = filename.lastIndexOf('.');
    return filename.slice(0, pointExtensionIndex);
}
exports.removeExtension = removeExtension;
function getBaseFilePath(filePath) {
    return "./" + path.basename(filePath);
}
exports.getBaseFilePath = getBaseFilePath;
function getClassName(fileContent) {
    return fileContent.match(/export\s*class\s*(\S*)\s*.*{/m)[1];
}
exports.getClassName = getClassName;
function matchAllEntities(inputString, regExp) {
    var tmp;
    var result = [];
    while ((tmp = regExp.exec(inputString)) !== null) {
        if (tmp.index === regExp.lastIndex) {
            regExp.lastIndex++;
        }
        result.push(tmp);
    }
    return result;
}
exports.matchAllEntities = matchAllEntities;
