// angular-unit-test-generator/constants/index-template-variables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var IndexTemplateVariables = (function () {
    function IndexTemplateVariables(opt, entityName, entityType, entityPath, spyDeclarations, spyArguments, spyInitializations, spyImports, methodsTests, spyUtilsPath) {
        if (opt === void 0) { opt = {}; }
        if (entityName === void 0) { entityName = opt.entityName || null; }
        if (entityType === void 0) { entityType = opt.entityType || null; }
        if (entityPath === void 0) { entityPath = opt.entityPath || null; }
        if (spyDeclarations === void 0) { spyDeclarations = opt.spyDeclarations || null; }
        if (spyArguments === void 0) { spyArguments = opt.spyArguments || null; }
        if (spyInitializations === void 0) { spyInitializations = opt.spyInitializations || null; }
        if (spyImports === void 0) { spyImports = opt.spyImports || null; }
        if (methodsTests === void 0) { methodsTests = opt.methodsTests || ''; }
        if (spyUtilsPath === void 0) { spyUtilsPath = opt.spyUtilsPath || null; }
        this.entityName = entityName;
        this.entityType = entityType;
        this.entityPath = entityPath;
        this.spyDeclarations = spyDeclarations;
        this.spyArguments = spyArguments;
        this.spyInitializations = spyInitializations;
        this.spyImports = spyImports;
        this.methodsTests = methodsTests;
        this.spyUtilsPath = spyUtilsPath;
    }
    return IndexTemplateVariables;
}());
exports.IndexTemplateVariables = IndexTemplateVariables;
